runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/WriteFilesTranslation.java

```diff
@@ -0,0 +1,152 @@
+
+  /** The URN for an unknown Java {@link FileBasedSink}. */
+  public static final String CUSTOM_JAVA_FILE_BASED_SINK_URN =
+      "urn:beam:file_based_sink:javasdk:0.1";
+
+  @VisibleForTesting
+  static WriteFilesPayload toProto(WriteFiles<?> transform) {
+    return WriteFilesPayload.newBuilder()
+        .setSink(toProto(transform.getSink()))
+        .setWindowedWrites(transform.isWindowedWrites())
+        .setRunnerDeterminedSharding(
+            transform.getNumShards() == null && transform.getSharding() == null)
+        .build();
+  }
+
+  private static SdkFunctionSpec toProto(FileBasedSink<?> sink) {
+    return SdkFunctionSpec.newBuilder()
+        .setSpec(
+            FunctionSpec.newBuilder()
+                .setUrn(CUSTOM_JAVA_FILE_BASED_SINK_URN)
+                .setParameter(
+                    Any.pack(
+                        BytesValue.newBuilder()
+                            .setValue(
+                                ByteString.copyFrom(SerializableUtils.serializeToByteArray(sink)))
+                            .build())))
+        .build();
+  }
+
+  @VisibleForTesting
+  static FileBasedSink<?> sinkFromProto(SdkFunctionSpec sinkProto) throws IOException {
+    checkArgument(
+        sinkProto.getSpec().getUrn().equals(CUSTOM_JAVA_FILE_BASED_SINK_URN),
+        "Cannot extract %s instance from %s with URN %s",
+        FileBasedSink.class.getSimpleName(),
+        FunctionSpec.class.getSimpleName(),
+        sinkProto.getSpec().getUrn());
+
+    byte[] serializedSink =
+        sinkProto.getSpec().getParameter().unpack(BytesValue.class).getValue().toByteArray();
+
+    return (FileBasedSink<?>)
+        SerializableUtils.deserializeFromByteArray(
+            serializedSink, FileBasedSink.class.getSimpleName());
+  }
+
+  public static <T> FileBasedSink<T> getSink(
+      AppliedPTransform<PCollection<T>, PDone, ? extends PTransform<PCollection<T>, PDone>>
+          transform)
+      throws IOException {
+    return (FileBasedSink<T>) sinkFromProto(getWriteFilesPayload(transform).getSink());
+  }
+
+  public static <T> boolean isWindowedWrites(
+      AppliedPTransform<PCollection<T>, PDone, ? extends PTransform<PCollection<T>, PDone>>
+          transform)
+      throws IOException {
+    return getWriteFilesPayload(transform).getWindowedWrites();
+  }
+
+  public static <T> boolean isRunnerDeterminedSharding(
+      AppliedPTransform<PCollection<T>, PDone, ? extends PTransform<PCollection<T>, PDone>>
+          transform)
+      throws IOException {
+    return getWriteFilesPayload(transform).getRunnerDeterminedSharding();
+  }
+
+  private static <T> WriteFilesPayload getWriteFilesPayload(
+      AppliedPTransform<PCollection<T>, PDone, ? extends PTransform<PCollection<T>, PDone>>
+          transform)
+      throws IOException {
+    return PTransformTranslation.toProto(
+            transform, Collections.<AppliedPTransform<?, ?, ?>>emptyList(), SdkComponents.create())
```

Comment:
I don't like having to provide the `componentTransforms` for translations at these points. We should consider an alias.